### DISCUSSION TITLE
How do I snap selection to edges/corners/etc?

### DISCUSSION THREAD

**Discussion by MatJab94** [tannerhelland/PhotoDemon#539] [Q&A]
With Rectangular Selection tool I expected to be able to snap selection to corners/edges of the image (and some other common points, like 1/2 of the size, 1/4 etc, in either axis). I can't figure out a way to do it.

I would've expected it to either:
- use snapping by default and be able to temporarily disable it (by holding Ctrl or whatever) while creating the selection
- not use snapping by default, but be able to temporarily enable it (by holding Ctrl or whatever) while creating the selection

I tried various combinations of Ctrl/Shift/Alt and could only get it to make a square selection if holding Shift.

**Comment by tannerhelland** [comment]
Hello @MatJab94 .  I apologize that PhotoDemon does not provide a "snap" option yet.  I originally planned to implement this alongside broader grid and guide support (which are still a ways out), but I could always move the snap option up the priority queue.

I usually refer to Photoshop and GIMP when adding new features.  Can you recommend any other software where you really like the snap selection behavior?  Additional references are helpful for me to understand the best way to approach a new feature.  Thank you in advance!

**Reply by MatJab94** [reply]
Alright, just wanted to make sure that I'm not missing anything obvious - if it's not implemented yet then I'll just have to wait. At least there's a way to manually type coordinates/dimensions for the selection, so that can be used as a (slow) workaround.

As for references - I mostly used Photoshop and I'm just used to how it works in general. But I'm no graphic designer or anything, for the most part I just need simple stuff like cropping/resizing/color adjustments, and from slightly more advanced features - macros to automate common tasks when I need to edit many images in bulk in the same way etc.

For my simple needs Photoshop's way too expensive (and I was fed up with how slow it is), so I've started looking around for alternatives. Haven't tried that many yet, but so far I'm kind of terrified of how bad the situation is - everything's some combination of slow, buggy, lacking features and is unintuitive/clunky/hard to switch to from Photoshop's way of doing things. And I thought Photoshop's bad...

I'll probably continue my search, because even though PhotoDemon looked very promising (and there's a couple of features I really liked), I still found some issues that make me hesitant to switch to it:
- no snapping features, like already mentioned
- the application starts pretty fast, but I can't smoothly move the view/use the tools - feels like it's running at ~20-30fps
- tried the macros feature and it seems like it's a bit buggy currently - I made a selection, typed in manual adjustements to its position/size, continued to do some other actions, and when I saved the macro and run it - it didn't properly re-create the selection (looks like it ignored the manual adjustments I typed in)
- on the topic of typing in size/coordinates of the selection, it's great that I can use math expressions in text fields, but I'd expect that pressing Enter would apply/deactivate the text field - instead I have to click away somewhere

**Reply by tannerhelland** [reply]
Thank you so much for this detailed feedback, @MatJab94 .  It is very helpful to hear detailed user experience expectations and patterns from a new user!

As you've seen, small developers struggle to compete with Photoshop.  We all kind of fixate on the little photo editing areas where we feel most competent, while struggling to bring other aspects of our apps up to Adobe's level.  (Competing with a multi-billion dollar corporation with 30,000 employees is tough!)  

In my case, I've largely fixated on making PhotoDemon as resource-lite as possible, while maintaining support all the way back to Windows XP for as long as I can.  This has made it increasingly challenging to improve performance on modern PCs, and striving for a balance between "supporting everyone's PC" and "not becoming so slow as to be unusable" remains an elusive goal.

I will do my best to solve some of the UI limitations you've mentioned in the coming days.  It matters a lot to me that small annoyances ("papercuts") are dealt with swiftly.  I also hope to add snapping support soon.

Wherever you end up in your app journey, thank you again for the feedback on PhotoDemon.  User feedback really does help small solo developers like me a lot!

**Comment by tannerhelland** [comment]
Just a quick FYI that snapping support is now available in nightly builds.  More improvements to come soon, I hope.
